File: lab1/Лаб 1 go/queue.go
package main

type QNode struct {
	Data string
	Next *QNode
}

type Queue struct {
	Front *QNode
	Rear  *QNode
}

func VectorToQueue(values []string) *Queue {
	q := &Queue{
		Front: nil,
		Rear:  nil,
	}
	for _, value := range values {
		newNode := &QNode{
			Data: value,
			Next: nil,
		}
		if q.Rear == nil {
			q.Front = newNode
			q.Rear = newNode
		} else {
			q.Rear.Next = newNode
			q.Rear = newNode
		}
	}
	return q
}

func QueueToVector(q *Queue) []string {
	values := []string{}
	current := q.Front
	for current != nil {
		values = append(values, current.Data)
		current = current.Next
	}
	return values
}

func QPush(values *[]string, value string) bool {
	q := VectorToQueue(*values)
	newNode := &QNode{
		Data: value,
		Next: nil,
	}
	if q.Rear == nil {
		q.Front = newNode
		q.Rear = newNode
	} else {
		q.Rear.Next = newNode
		q.Rear = newNode
	}
	*values = QueueToVector(q)
	fmt.Printf("Элемент '%s' добавлен в очередь\n", value)
	return true
}

func QPop(values *[]string) bool {
	q := VectorToQueue(*values)
	if q.Front == nil {
		fmt.Println("Ошибка: очередь пуста")
		return false
	}
	poppedValue := q.Front.Data
	q.Front = q.Front.Next
	if q.Front == nil {
		q.Rear = nil
	}
	*values = QueueToVector(q)
	fmt.Printf("Элемент '%s' удален из очереди\n", poppedValue)
	return true
}

File: lab1/Лаб 1/queue.cpp
#include "queue.h"
#include <iostream>
#include <vector>
#include <string>

using namespace std;

Queue* vectorToQueue(const vector<string>& values) {
    Queue* q = new Queue;
    q->front = nullptr;
    q->rear = nullptr;
    
    for (const string& value : values) {
        QNode* newNode = new QNode;
        newNode->data = value;
        newNode->next = nullptr;
        
        if (q->rear == nullptr) {
            q->front = newNode;
            q->rear = newNode;
        } else {
            q->rear->next = newNode;
            q->rear = newNode;
        }
    }
    
    return q;
}

void queueToVector(Queue* q, vector<string>& values) {
    values.clear();
    
    QNode* current = q->front;
    while (current != nullptr) {
        values.push_back(current->data);
        current = current->next;
    }
}

void freeQueue(Queue* q) {
    QNode* current = q->front;
    while (current != nullptr) {
        QNode* next = current->next;
        delete current;
        current = next;
    }
    delete q;
}

bool qpush(vector<string>& values, const string& value) {
    Queue* q = vectorToQueue(values);
    
    QNode* newNode = new QNode;
    newNode->data = value;
    newNode->next = nullptr;
    
    if (q->rear == nullptr) {
        q->front = newNode;
        q->rear = newNode;
    } else {
        q->rear->next = newNode;
        q->rear = newNode;
    }
    
    queueToVector(q, values);
    freeQueue(q);
    
    cout << "Элемент '" << value << "' добавлен в очередь" << endl;
    return true;
}

bool qpop(vector<string>& values) {
    Queue* q = vectorToQueue(values);
    
    if (q->front == nullptr) {
        cerr << "Ошибка: очередь пуста" << endl;
        freeQueue(q);
        return false;
    }
    
    string poppedValue = q->front->data;
    QNode* nodeToDelete = q->front;
    q->front = q->front->next;
    
    if (q->front == nullptr) {
        q->rear = nullptr;
    }
    
    delete nodeToDelete;
    
    queueToVector(q, values);
    freeQueue(q);
    
    cout << "Элемент '" << poppedValue << "' удален из очереди" << endl;
    return true;
}